x/programs/examples/imports/pstate/pstate.go

```diff
@@ -217,9 +217,45 @@ func (i *Import) deleteFn(caller *program.Caller, memOffset int32, size int32) (
 	}
 
 	k := storage.ProgramPrefixKey(args.ProgramID[:], args.Key)
+	bytes, err = i.mu.GetValue(context.Background(), k)
+	if err != nil {
+		if errors.Is(err, database.ErrNotFound) {
+			val, err := program.WriteBytes(memory, []byte{0})
+			if err != nil {
+				i.log.Error("failed to write to memory",
+					zap.Error(err),
+				)
+				return nil, err
+			}
+
+			return types.ValI32(int32(val)), nil
+		}
+
+		i.log.Error("failed to get value from storage",
+			zap.Error(err),
+		)
+		return nil, err
+	}
+
+	// prepend 1 to val
+	bytes = append([]byte{1}, bytes...)
+
+	ptr, err := program.WriteBytes(memory, bytes)
+	if err != nil {
+		{
+			i.log.Error("failed to write to memory",
+				zap.Error(err),
+			)
+		}
+		return nil, err
+	}
+
 	if err := i.mu.Remove(context.Background(), k); err != nil {
-		i.log.Error("failed to remove from storage", zap.Error(err))
-		return types.ValI32(-1), nil
+		i.log.Error("failed to delete value from storage",
+			zap.Error(err),
+		)
+		return nil, err
```

Comment:
Since you have set the new pointer type to a `NonNull<_>`, returning a `nil` value would violate the property of this type: `Unlike *mut T, the pointer must always be non-null, even if the pointer is never dereferenced. This is so that enums may use this forbidden value as a discriminant – Option<NonNull<T>> has the same size as *mut T. However the pointer may still dangle if it isn’t dereferenced.` but I just wanted to make sure that returning a non-nil error from these linked functions would indeed cause a panic in the host, it seems to be what is happening in wasmtime https://pkg.go.dev/github.com/bytecodealliance/wasmtime-go#NewFunc can you please confirm my understanding ?

Reply:
I believe our linters force us to handle all errors. If they don't they should. There are a number of places in the this code that return the `nil, err` tuple. I'm just following the convention that was already set up